### PR TITLE
[#22737] Fit & Finish for Go SDK timer support.

### PR DESCRIPTION
**Pull request by lostluck**
Cleanup and fit & finish work for Go SDK timers. 

* Adds a timers.Context for providing the Family & Tag fields.
* Adds a ClearTag method to EventTime and ProcessingTime timers to support clearing a timer for a specific tag.
* Updates the timer_wordcap example
  * Use a finite Periodic sequence to get elements.
  * Demonstrate element batching with State & Timers.
  * Add verbose logging to clarify execution for example purposes.
  * Demonstrate multiple tags set for the same timer.
* Simplifies user key coderID sourcing by moving it entirely to pipeline serialization time.
  * Current Beam model is more general on timer key and window coders, but they are implicitly associated directly with the DoFn's parallel input key.
* Fetches coders at execution time from the bundle descriptors returned timer_family_spec.
  * This avoids relying on the previous implementation detail, while also allowing the runner to replace the user key coder as needed, such as to add length prefixes or similar.
* Handle inserted timers similarly to [Java SDK](https://github.com/apache/beam/blob/f154d0165d31068f9e0d2ae008d7823dcf4cf53a/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java#L1667-L1734) allowing for quickly re-looping timers when applicable.
  * Supporting this is the bulk of the PRs complexity, but ordering is required for correctness and avoiding double firing.
* Ensure keys are not repeatedly encoded/decoded for a given inline process.
  * Key bytes on fired timers are re-used when writing new timers. 
* Removes unnecessary UserTimerAdapter interface, as all uses are in the exec package.
* Ensures all timers from a given timer batch are parsed.
* Since they don't have mutable state, made timers.ProcessingTime and timers.EventTime have value receivers instead of pointer receivers.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @riteshghorse 

I'm aware you have Release things to do, so no rush for this. I still need to add unit tests for the processTimer method loop. But otherwise this should get timers to the finish line.

**Comment by lostluck** [comment]
Run Go PostCommit

**Comment by lostluck** [comment]
@riteshghorse this should now be ready for a look, tested and all.

**Comment by lostluck** [comment]
Release definitely comes first. :)

File: sdks/go/pkg/beam/core/runtime/exec/pardo.go [hunk 1]
```diff
@@ -362,47 +370,128 @@ func (n *ParDo) invokeDataFn(ctx context.Context, pn typex.PaneInfo, ws []typex.
...
+
+	bundleTimers, err := decodeBundleTimers(spec, r)
 	if err != nil {
-		return nil, errors.WithContext(err, "error decoding received timer callback")
+		return err
 	}
+	for _, tmap := range bundleTimers {
+		n.TimerTracker.SetCurrentKeyString(tmap.KeyString)
+		for i, w := range tmap.Windows {
+			ws := tmap.Windows[i : i+1]
```
Comment (riteshghorse) [inline sdks/go/pkg/beam/core/runtime/exec/pardo.go:413]: is it okay to use `w` here instead of `tmap.Windows[i : i+1]`

**Review by riteshghorse — COMMENTED** [review]
Great work again @lostluck 
Mostly amazed by how things are handled in exec/timer.go. Small nits and questions but looks good.

**Comment by lostluck** [comment]
Looks like Jenkins is broken outside for reasons unrelated to this PR.

File: sdks/go/pkg/beam/core/runtime/exec/pardo.go (hunk 1 above)
Comment (riteshghorse) [reply sdks/go/pkg/beam/core/runtime/exec/pardo.go:413]: got it

File: sdks/go/pkg/beam/core/runtime/exec/pardo.go [hunk 2]
```diff
@@ -249,6 +253,10 @@ func (n *ParDo) FinishBundle(_ context.Context) error {
 	if _, err := n.invokeDataFn(n.ctx, typex.NoFiringPane(), window.SingleGlobalWindow, mtime.ZeroTimestamp, n.Fn.FinishBundleFn(), nil); err != nil {
 		return n.fail(err)
 	}
+	// Flush timers if any.
+	if err := n.TimerTracker.FlushAndReset(n.ctx, n.timerManager); err != nil {
```
[the remark this answers is not in the document]
Comment (riteshghorse) [reply sdks/go/pkg/beam/core/runtime/exec/pardo.go:257]: okay

**Review by riteshghorse — APPROVED** [review]
LGTM, thanks!

**Comment by lostluck** [comment]
Run RAT PreCommit

**Comment by lostluck** [comment]
Run RAT PreCommit

**Comment by lostluck** [comment]
Run Go PreCommit

**Comment by lostluck** [comment]
Run GoPortable PreCommit

**Comment by lostluck** [comment]
Run RAT PreCommit

**Pull request merged**